vochain/state/balances.go

```diff
@@ -221,43 +222,82 @@ func (v *State) ConsumeFaucetPayload(from common.Address, faucetPayload *models.
 // TransferBalance transfers balance from origin address to destination address,
 // and updates the state with the new values (including nonce).
 // If origin address acc is not enough, ErrNotEnoughBalance is returned.
-func (v *State) TransferBalance(from, to common.Address, amount uint64) error {
-	accFrom, err := v.GetAccount(from, false)
+func (v *State) TransferBalance(tx *vochaintx.TokenTransfer, burnTxCost bool) error {
+	accFrom, err := v.GetAccount(tx.FromAddress, false)
 	if err != nil {
 		return err
 	}
 	if accFrom == nil {
 		return ErrAccountNotExist
 	}
-	accTo, err := v.GetAccount(to, false)
+	accTo, err := v.GetAccount(tx.ToAddress, false)
 	if err != nil {
 		return err
 	}
 	if accTo == nil {
 		return ErrAccountNotExist
 	}
-	if err := accFrom.Transfer(accTo, amount); err != nil {
+	if err := accFrom.Transfer(accTo, tx.Amount); err != nil {
 		return err
 	}
-	log.Debugf("transferring %d tokens from %s to %s", amount, from.String(), to.String())
-	if err := v.SetAccount(from, accFrom); err != nil {
+	log.Debugf("transferring %d tokens from %s to %s", tx.Amount, tx.FromAddress.String(), tx.ToAddress.String())
+	if err := v.SetAccount(tx.FromAddress, accFrom); err != nil {
 		return err
 	}
-	if err := v.SetAccount(to, accTo); err != nil {
+	if err := v.SetAccount(tx.ToAddress, accTo); err != nil {
 		return err
 	}
-	for _, l := range v.eventListeners {
-		l.OnTransferTokens(from.Bytes(), to.Bytes(), amount)
+	if !burnTxCost {
+		for _, l := range v.eventListeners {
+			l.OnTransferTokens(&vochaintx.TokenTransfer{
+				FromAddress: tx.FromAddress,
+				ToAddress:   tx.ToAddress,
+				Amount:      tx.Amount,
+				TxHash:      tx.TxHash,
+				Height:      uint64(v.CurrentHeight()),
+			})
+		}
 	}
 	return nil
 }
 
 // MintBalance increments the existing acc of address by amount
-func (v *State) MintBalance(address common.Address, amount uint64) error {
-	if amount == 0 {
+func (v *State) MintBalance(tx *vochaintx.TokenTransfer) error {
+	if tx.Amount == 0 {
 		return fmt.Errorf("cannot mint a zero amount balance")
 	}
-	acc, err := v.GetAccount(address, false)
+	acc, err := v.GetAccount(tx.ToAddress, false)
+	if err != nil {
+		return fmt.Errorf("mintBalance: %w", err)
+	}
+	if acc == nil {
+		return ErrAccountNotExist
+	}
+	if acc.Balance+tx.Amount <= acc.Balance {
```

Comment:
nit: you want `<` to actually catch an overflow, otherwise this will also trigger if `tx.Amount == 0`.

Reply:
(I think the existing code also has this potential flaw)

Reply:
At this point, we've already checked that tx.Amount is not 0, so I cannot see when this case can happen.

Reply:
just a matter of defensive programming :) even if we can be sure that today Amount==0 is impossible, it might be possible in the future after refactors or we might have got it wrong. up to you of course, I just prefer to write defensive code in general.

Reply:
fair enough, added the suggested change :) Thank you dani!